src/mui/input/FileInput.js

```diff
@@ -0,0 +1,139 @@
+import React, { Component, PropTypes } from 'react';
+import Dropzone from 'react-dropzone';
+import CircularProgress from 'material-ui/CircularProgress';
+
+const defaultStyle = {
+    dropZone: {
+        background: '#efefef',
+        cursor: 'pointer',
+        padding: '1rem',
+        textAlign: 'center',
+        color: '#999',
+    },
+    previewContainer: {
+        position: 'relative',
+        display: 'inline-block',
+    },
+    preview: {
+        margin: '.5rem',
+        maxHeight: '10rem',
+    },
+    loader: {
+        margin: 0,
+        display: 'block',
+        position: 'absolute',
+        left: '50%',
+        top: '50%',
+        transform: 'translate(-50%, -50%)',
+    },
+    uploading: {
+        opacity: .3,
+    }
+};
+
+class FileInput extends Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            files: [],
+        };
+    }
+
+    onDrop = files => {
+        const loadingFiles = files.map(f => ({
+            data: f, // destructuring a File object remove almost all its properties
+            uploading: true,
+        }));
+        this.setState({ files: loadingFiles });
+
+        this.props.onUpload(files, () => {
+            const loadedFiles = files.map(f => ({
+                data: f,
+                uploading: false,
+            }));
+            this.setState({ files: loadedFiles });
+        });
+    }
+
+    label() {
+        if (this.props.multiple) {
+            return <p>Drop some files to upload, or click to select one.</p>;
+        }
+
+        return <p>Drop a file to upload, or click to select it.</p>;
+    }
+
+    renderFile = file => {
```

Comment:
I'm not a fan of the renderXXX method. Can't you use a separate component here instead ?